File: 409_Longest_Palindrome.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/longest-palindrome/
# Given a string which consists of lowercase or uppercase letters, find the length of the longest palindromes that
# can be built with those letters. This is case sensitive, for example "Aa" is not considered a palindrome here.

# Iterate over s maintaining a set of letters seen. If a letter is seen again, add 2 to length of palindrome and
# delete from set. If a letter is not in set, add it. Add 1 for a single letter in the middle of the palindrome if
# there are any single letters after iteration.
# Time - O(n)
# Space - O(1)

class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: int
        """
        max_length = 0
        letters = set()

        for c in s:
            if c in letters:
                max_length += 2
                letters.remove(c)
            else:
                letters.add(c)

        return max_length + bool(letters)

File: java/409_Longest_Palindrome.java
/*
https://leetcode.com/problems/longest-palindrome/
Given a string which consists of lowercase or uppercase letters, find the length of the longest palindromes that
can be built with those letters. This is case sensitive, for example "Aa" is not considered a palindrome here.

Create a set of all chars with odd frequenceies. When a char has even frequency, remove from set and add 2 to length.
If any odd frequency remains, add 1 to length for central char.
Time - O(n)
Space - O(1)
*/

public class Solution {
    public int longestPalindrome(String s) {

        HashSet<Character> odd = new HashSet<Character>();
        int longest = 0;

        for (int i = 0; i < s.length(); ++i) {
            char c = s.charAt(i);
            if (odd.contains(c)) {
                odd.remove(c);
                longest += 2;
            }
            else
                odd.add(c);
        }

        if (!odd.isEmpty())
            ++longest;
        return longest;
    }
}
